docs/interfaces.md

```diff
@@ -0,0 +1,568 @@
+// ECC_Enclave and TLCC_Enclave. Once a secure channel is established,
+// Session Requests to TLCC can be issued to invoke tlcc functions such
+// as getMetadata or verifyMSPId.
+
+// TODO How to integrate the the message authenticator?!
+// should we abstract it as signature here even though we may
+// use cmac once a session is established?
+
+message SessionSetup {
+        // TBD input enclave vk?
+        // TLCC Mrenclave?
+        string channel_id = 3;
+}
+
+message SessionSetupResponse {
+        SessionSetup setup = 1;
+        byte context       = 2;         // context that references the session
+}
+
+message SessionClose {
+        byte context = 1;
+}
+
+message SessionCloseResponse {
+}
+
+// TODO add signature or some othere for of message
+// authentication to session request/response.
+
+message SessionRequest {
+        byte context    = 1;
+        byte tx_context = 2;
+        byte nonce      = 3;
+        oneof request {
+                GetMetadataRequest metadata     = 1; 
+                validateMSPRequest validate_msp = 2;
+        }
+}
+
+message SessionRequeatResponse {
+        SessionRequest request = 1;
+        oneof response {
+                GetMetadataResponse metadata     = 1; 
+                validateMSPResponse validate_msp = 2; 
+        }
+}
+
+```
+
+TLCC keeps a local reference of the enclave id. This reference is volatile. Also a reference to the peer is kept to access the ledger and read blocks.
+
+```go
+type TrustedLedgerCC struct {
+	enclave enclave.Stub
+	peer    *peer.Peer
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+# TLCC_Enclave
+
+## Interface:
+Defined in the TLCC EDL.
+```c++
+// creates a new channel
+public int ecall_join_channel(
+        [in, size=gen_len] uint8_t *genesis, uint32_t gen_len);
+
+// pushes next block to tlcc
+public int ecall_next_block(
+        [user_check] uint8_t *block_bytes, uint32_t block_size);
+```
+
+```c++
+// this secure session functions are defined in a secure channel module
+//included from secure_channel.edl
+
+public int session_setup(
+        [in] setup SessionSetup,
+        [out] SessionSetupResponse);
+
+public int session_close(
+        [in] close SessionClose
+        [out] SessionCloseReponse);
+
+public int session_request(
+        [in] request SessionRequest,
+        [out] SessionRequestResponse;
+```
+
+```c++
+// addtionally we may use the state sealing module to persists TLCC state
+// see below the data sealing module.
+include sealed_storage.edl
+```
+
+Session requests are depatched to the following interfaces:
+```c++
+// TODO refine
+// verify state data
+public uint metadata get_state_metadata(
+        const char *namespace,
+        const char *key);
+
+public int get_multi_state_metadata(
+        [in, string] const char *namespace,
+        [in, string] const char *comp_key,
+        [out] uint8_t *metadata);
+
+// verify msp identities
+public bool validate_msp_id(
```

Comment:
I changed this function to validate_identity, which can be used to implement get_creator and check if the creator is a valid identity with respect to the consortium defined in this channel.

Reply:
renamed it to validate_identity, which can be used to implement getCreator

Reply:
Ok, it should be enough for now.

However, this [document (part 2)](https://hyperledger-fabric.readthedocs.io/en/release-2.2/txflow.html) corroborates my second point: we need a mechanism for checking the proposal (including its creator, and thus this API).

Reply:
we definitely need to validate the signed proposal.  Using the proposed `validate_identity` we should be able that the creator -- as expressed by `serializedIdentity` which contains the msp-id and the x509 cert with pk, name and role [might help, though, explicitly mention that in the comment around `validate_identity` -- is a valid party grounded in the specified MSP-id root-of-trust and the rest of the signature verification of the proposal we can do locally to ecc based on the resulted validated public key?  Is there any other step in verifying the signedproposal in the enclave we need `tlcc` assistance?

Reply:
(see the link I have attached)
The endorsing peer verifies also that (i) the proposal has not been submitted in the past, (ii) the submitter respects the channel policy. In theory, TLCC should do this. In practice, we can rely on the validators for (i), and delegate (ii, or more precisely, any authentication/authorization) to the enclave.

Reply:
ok, i see now what you mean (didn't initially get what you meant by "part 2").  Good points.

Regarding the replay protection, at first sight it certainly seems mainly a waste of resource you save by checking, as during submission of a replayed transaction it should fail due to rw-set conflicts?  This assumes, though, it is state-changing, maybe there could be issues of transactions which might be non-state-changing dependent on state and somebody could try to replay such a transaction at a later time when it would be state-changing? So integrity-wise there might or might not be issue?  For FPC maybe confidentiality could add additional twists: As an original tx/query submitter can always recreate a new query, so from overt channel (results to querier) perspective a replay shouldn't help?  If there would be covert channel, then a resubmit could potentially help in throttling attacks but that seems rather esoteric?  So i would somewhat agree that we can punt on replay protection to validators/rw-conflict (although i have not a good feeling yet how much this _this-tx-does-not-change-state-now-but-could-potentially-later scenario_ is an issue or not.)

regarding authorization, though, it becomes a bit more sensitive as we are talk also about confidentiality (for queries), not only integrity.  So maybe the `validated_identity` should maybe return which permissions the client has and ecc should immediately abort if client doesn't have `read` privs and abort on first write-attempt if it doesn't have `write` permissions?  Of course, we could also punt on that but would have to document that reader/writer polices could not be trusted to provide confidentiality of information protected fpc cc?

Reply:
> during submission of a replayed transaction it should fail due to rw-set conflicts?

**Probably yes**, in a concurrent execution: send proposal P1 and replay it immediately; you get two separate endorsements/transactions where the updated keys likely conflict with each other.
**Maybe not**, in a serialized execution: send proposal P1 (e.g., incrementCounter), commit transaction; replay P1, commit transaction; here the counter will be incremented by two with a single proposal.

The check that Fabric does should be [here](https://github.com/hyperledger/fabric/blob/c799dcc188d76c5d3f3e2cdab931f30a00273d17/core/endorser/endorser.go#L275). It's based a transaction id value computed over the proposal and stored (at some point) on the ledger.

> So maybe the validated_identity should maybe return which permissions the client has and ecc should immediately abort if client doesn't have read privs and abort on first write-attempt if it doesn't have write permissions?

In theory, yes.
However, this is all Fabric logic that TLCC should implement. Following Yacov's recommendation, we can simplify.

> Of course, we could also punt on that but would have to document that reader/writer polices could not be trusted to provide confidentiality of information protected fpc cc?

In practice, this is (part of) the solution. The other part (in my answer above) is to tell a developer how to leverage the fpc chaincode to implement an equivalent mechanism.

Reply:
> > during submission of a replayed transaction it should fail due to rw-set conflicts?
> [..]
> **Maybe not**, in a serialized execution: send proposal P1 (e.g., incrementCounter), commit transaction; replay P1, commit transaction; here the counter will be incremented by two with a single proposal.

Oops, of course, this is a much simpler examples where replay works (badly) than my contrieved sometimes-state-changes example. 

> The check that Fabric does should be [here](https://github.com/hyperledger/fabric/blob/c799dcc188d76c5d3f3e2cdab931f30a00273d17/core/endorser/endorser.go#L275). It's based a transaction id value computed over the proposal and stored (at some point) on the ledger.

This code seems to be executed by the endorsers. You mentioned above also "rely on the validators for (i)": Do the validators automatically also cross-check this?  If so, i think we can punt on them as good peers would ignore replays from a integrity perspective and from a confidentiality perspective there seems only esoteric direct leakage (and not indirect via integrity corruption, which would be caught by validators?)   If standard validators would not do it, this (contrary to authorizaton) seems a bit a tough to say we do not support replay protetion and  punt it to the app developer?   If we can live with some false-positives for replays, we could handle replay protection with a bloom-filter.  Actually, for channel & ercc transactions we cannot punt to app developers at all? So we more or less have something inside tlcc?   We currently have these three opaque kvs_t files as part of tlcc internal state but seems we might want to be more explicitl. And of course this ties now this to (some) resolutions of #402 ... :-(

> > So maybe the validated_identity should maybe return which permissions the client has and ecc should immediately abort if client doesn't have read privs and abort on first write-attempt if it doesn't have write permissions?
> 
> In theory, yes.
> However, this is all Fabric logic that TLCC should implement. Following Yacov's recommendation, we can simplify.

That's what i meant with punt below. But i agree that besides explicit telling you cannot rely on ledger-specified authorization rules (which we would have to do at a minimum) ...
 
> > Of course, we could also punt on that but would have to document that reader/writer polices could not be trusted to provide confidentiality of information protected fpc cc?
> 
> In practice, this is (part of) the solution. The other part (in my answer above) is to tell a developer how to leverage the fpc chaincode to implement an equivalent mechanism.

... pointing out how to do it themselves is even better.  Certainly, per-app authorization policies should be easy to handle via `get_creator_name` ...